src/app/navigation/utils.ts

```diff
@@ -109,3 +112,40 @@ export const findPathForParams = (params: object, paths: string[]) => {
       paramsInPath.every(({name}) => paramKeys.includes(name.toString()));
   });
 };
+
+const isScrollTarget = (
+  object: { [key: string]: any }
+): object is HighlightScrollTarget | SearchScrollTarget => {
+  if (typeof object.elementId !== 'string') { return false; }
+  if (object.type === 'search' && typeof object.index === 'number') { return true; }
+  if (object.type === 'highlight' && typeof object.id === 'string') { return true; }
+  return false;
+};
+
+export const getScrollTargetFromQuery = (
+  query: OutputParams,
+  hash: string
+): HighlightScrollTarget | SearchScrollTarget | null => {
+  if (!hash || !query.target || Array.isArray(query.target)) { return null; }
+  try {
+    const parsed = JSON.parse(decodeURIComponent(query.target));
+    if (parsed instanceof Object) {
+      parsed.elementId = hash.replace('#', '');
+      if (isScrollTarget(parsed)) { return parsed; }
+    }
+    return null;
+  } catch {
```

Comment:
it'd be nice if this catch was only around the JSON.parse or limited to only the deserialization error